File: face/geomajas-face-dojo/geomajas-dojo-client/src/main/resources/geomajas/spatial/Bbox.js
dojo.provide("geomajas.spatial.Bbox");
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
dojo.require("geomajas.spatial.Coordinate");

dojo.declare("Bbox", null, {

	/**
	 * @fileoverview Bounding Box.
	 * @class Bounding Box.<br>
	 * Determined by a point, it's width and it's height. If the 4 params
	 * aren't correct, everything is set to 0.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 * @param x The x-value for the origin-point.
	 * @param y The y-value for the origin-point.
	 * @param width The bounding box' width (along the X-axis).
	 * @param height The bounding box' height (along the Y-axis).
	 */
	constructor : function (x, y, width, height) {
		if (x != null && y != null && width != null && height != null) {
			this.x = x;
			this.y = y;
			this.width = width;
			this.height = height;
		} else {
			this.x = 0;
			this.y = 0;
			this.width = 0;
			this.height = 0;
		}
	},

	toJSON : function () {
		return {
			javaClass : "org.geomajas.geometry.Bbox",
			x : this.x,
			y : this.y,
			width : this.width,
			height : this.height
		};
	},

	/**
	 * Set new values for this Bbox.
	 * @param x The x-value for the origin-point.
	 * @param y The y-value for the origin-point.
	 * @param width The bounding box' width (along the X-axis).
	 * @param height The bounding box' height (along the Y-axis).
	 */
	set : function (/*Double*/x, /*Double*/y, /*Double*/width, /*Double*/height) {
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
	},

	equals : function (bbox) {
		return (
			this.x == bbox.x &&
			this.y == bbox.y &&
			this.width == bbox.width &&
			this.height == bbox.height
		);
	},

	clone : function () {
		return new Bbox(this.x, this.y, this.width, this.height);
	},
	
	/**
	 * Returns the origin (x, y) as a Coordinate.
	 */
	getOrigin : function () {
		return new Coordinate (this.x, this.y);
	},
	
	/**
	 * Clones and resizes to the specified width and height, keeping the center fixed
	 */
	cloneWithSize : function (width, height) {
		var clone = this.clone();
		clone.setWidth(width);
		clone.setHeight(height);
		clone.setCenterPoint(this.getCenterPoint());
		return clone;
	},
	
	/**
	 * Moves center to the specified coordinate
	 */
	setCenterPoint : function (coordinate) {
		this.x = coordinate.getX() - 0.5 * this.width;
		this.y = coordinate.getY() - 0.5 * this.height;
	},
	
	/**
	 * Get the center of the bounding box as a Coordinate.
	 */
	getCenterPoint : function () {
		return new Coordinate (this.x + this.width/2, this.y + this.height/2);
	},

	/**
	 * Get the endpoint of the bounding box as a Coordinate.
	 */
	getEndPoint : function () {
		return new Coordinate (this.x + this.width, this.y + this.height);
	},

	/**
	 * Get the coordinates of the bounding box (5 coordinates) as an array.
	 */
	getCoordinates : function () {
		var result = [];
		result.push(new Coordinate(this.x,this.y));
		result.push(new Coordinate(this.x+ this.width,this.y));
		result.push(new Coordinate(this.x+ this.width,this.y+ this.height));
		result.push(new Coordinate(this.x,this.y+ this.height));
		result.push(new Coordinate(this.x,this.y));
		return result;
	},

	/**
	 * Does this bounding box contain the given bounding box?
	 * @param other Another Bbox.
	 * @return true is the other is completely surrounded by this one,
	 *         false otherwise.
	 */
	contains : function (/*Bbox*/other) {
		if(other.getX() < this.getX()){
			return false;
		}
		if(other.getY() < this.getY()){
			return false;
		}
		if(other.getEndPoint().getX() > this.getEndPoint().getX()){
			return false;
		}
		if(other.getEndPoint().getY() > this.getEndPoint().getY()){
			return false;
		}
		return true;		
	},
	

	/**
	 * Does this bounding box contain the given coordinate ?
	 * @param other Another Bbox.
	 * @return true is the coordinate is completely surrounded by this bbox
	 */
	containsCoordinate : function (/*Coordinate*/coord) {
		if(coord.getX() < this.getX()){
			return false;
		}
		if(coord.getY() < this.getY()){
			return false;
		}
		if(coord.getX() > this.getEndPoint().getX()){
			return false;
		}
		if(coord.getY() > this.getEndPoint().getY()){
			return false;
		}
		return true;		
	},
	
	/**
	 * Does this bounding box intersect the given bounding box?
	 * @param other Another Bbox.
	 * @return true if the other intersects this one,
	 *         false otherwise.
	 */
	intersects : function (/*Bbox*/other) {
		//dojo.debug(other);
    	if(other.getX() > this.getEndPoint().getX()){
    		return false;
    	}
    	if(other.getY() > this.getEndPoint().getY()){
    		return false;
    	}
    	if(other.getEndPoint().getX() < this.getX()){
    		return false;
    	}
    	if(other.getEndPoint().getY() < this.getY()){
    		return false;
    	}
		return true;		
	},
	
	/**
	 * Computes the intersection of this bounding box with the specified bounding box
	 * @param other Another Bbox.
	 * @return bounding box of intersection
	 */
	intersection : function(/*Bbox*/other) {
		if(!this.intersects(other)){
			return new Bbox();
		} else {
			var minx = other.getX() > this.getX() ? other.getX() : this.getX();
			var maxx = other.getEndPoint().getX() < this.getEndPoint().getX() ? other.getEndPoint().getX() : this.getEndPoint().getX();
			var miny = other.getY() > this.getY() ? other.getY() : this.getY();
			var maxy = other.getEndPoint().getY() < this.getEndPoint().getY() ? other.getEndPoint().getY() : this.getEndPoint().getY();
			return new Bbox(minx,miny,(maxx-minx),(maxy-miny));
		}
	},

	/**
	 * Calculates the union of 2 bounding boxes.
	 * @param other The other Bbox.
	 */
	union : function (/*Bbox*/other) {
		if (other.getWidth() == 0 || other.getHeight == 0) {
			return this.clone();
		}
		if (this.width == 0 || this.height == 0) {
			return other.clone();
		}
		var minx = other.getX() < this.getX() ? other.getX() : this.getX();
		var maxx = other.getEndPoint().getX() > this.getEndPoint().getX() ? other.getEndPoint().getX() : this.getEndPoint().getX();
		var miny = other.getY() < this.getY() ? other.getY() : this.getY();
		var maxy = other.getEndPoint().getY() > this.getEndPoint().getY() ? other.getEndPoint().getY() : this.getEndPoint().getY();
		return new Bbox(minx,miny,(maxx-minx),(maxy-miny));
	},
	
	buffer : function (distance) {
		if (distance > 0) {
			return new Bbox(this.x - distance,this.y - distance, this.width+(distance*2),this.height+(distance*2));
		}
		return this.clone();
	},

	toString : function () {
		return "[x="+parseInt(this.x)+", y="+parseInt(this.y)+", w="+parseInt(this.width)+", h="+parseInt(this.height)+"]";
	},

	// Getters and setters.

	getX : function () {
		return this.x;
	},

	setX : function (/*Double*/x) {
		this.x = x;
	},

	getY : function () {
		return this.y;
	},

	setY : function (/*Double*/y) {
		this.y = y;
	},

	getOrigX : function () {
		return this.x;
	},

	getOrigY : function () {
		return this.y;
	},

	getEndX : function () {
		return this.x + this.width;
	},

	getEndY : function () {
		return this.y + this.height;
	},

	getWidth : function () {
		return this.width;
	},

	setWidth : function (/*Double*/width) {
		this.width = width;
	},

	getHeight : function () {
		return this.height;
	},

	setHeight : function (/*Double*/height) {
		this.height = height;
	},
	
	translate : function (/*Double*/dx, /*Double*/dy){
		this.x = this.x + dx;
		this.y = this.y + dy;
	}
});

File: face/geomajas-face-gwt/geomajas-gwt-client/src/main/java/org/geomajas/gwt/client/spatial/Bbox.java
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.geomajas.gwt.client.spatial;

import org.geomajas.geometry.Coordinate;

/**
 * <p>
 * Definition of an (axis aligned) Bounding Box. Determined by an x-ordinate an y-ordinate, it's width and it's height.
 * Notice the the name says "Axis Aligned"! This implies that this type of bounding box cannot rotate!
 * </p>
 * 
 * @author Pieter De Graef
 */
public class Bbox {

	/**
	 * The ordinate along the X-axis.
	 */
	private double x;

	/**
	 * The ordinate along the Y-axis.
	 */
	private double y;

	/**
	 * The bounding box' width.
	 */
	private double width;

	/**
	 * The bounding box' height.
	 */
	private double height;

	// -------------------------------------------------------------------------
	// Constructors:
	// -------------------------------------------------------------------------

	/**
	 * Default constructor setting the position to (0, 0), and a width and height of 1.
	 */
	public Bbox() {
		this(0, 0, 1, 1);
	}

	/**
	 * constructor that immediately applies values to all the fields.
	 * 
	 * @param x
	 *            The ordinate along the X-axis.
	 * @param y
	 *            The ordinate along the Y-axis.
	 * @param width
	 *            The bounding box' width.
	 * @param height
	 *            The bounding box' height.
	 */
	public Bbox(double x, double y, double width, double height) {
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
	}

	/**
	 * Construct a bbox by copying another one.
	 * 
	 * @param bounds
	 *            Another bounding box instance.
	 */
	public Bbox(Bbox bounds) {
		x = bounds.getX();
		y = bounds.getY();
		width = bounds.getWidth();
		height = bounds.getHeight();
	}

	/**
	 * Constructor that transforms the DTO bounding box to the GWT bounding box.
	 * 
	 * @param bounds
	 *            A DTO bounding box instance.
	 */
	public Bbox(org.geomajas.geometry.Bbox bounds) {
		x = bounds.getX();
		y = bounds.getY();
		width = bounds.getWidth();
		height = bounds.getHeight();
	}

	// -------------------------------------------------------------------------
	// Class specific functions:
	// -------------------------------------------------------------------------

	/**
	 * Create a clone of the object.
	 */
	public Object clone() {
		return new Bbox(x, y, width, height);
	}

	/**
	 * Return the origin (x, y) as a Coordinate.
	 */
	public Coordinate getOrigin() {
		return new Coordinate(x, y);
	}

	/**
	 * Get the center of the bounding box as a Coordinate.
	 */
	public Coordinate getCenterPoint() {
		double centerX = (width == 0 ? x : x + width / 2);
		double centerY = (height == 0 ? y : y + height / 2);
		return new Coordinate(centerX, centerY);
	}

	/**
	 * Get the end-point of the bounding box as a Coordinate.
	 */
	public Coordinate getEndPoint() {
		return new Coordinate(x + width, y + height);
	}

	/**
	 * Get the coordinates of the bounding box as an array.
	 * 
	 * @return Returns 5 coordinates so that the array is closed. This can be useful when using this array to creating a
	 *         <code>LinearRing</code>.
	 */
	public Coordinate[] getCoordinates() {
		Coordinate[] result = new Coordinate[5];

		result[0] = new Coordinate(x, y);
		result[1] = new Coordinate(x + width, y);
		result[2] = new Coordinate(x + width, y + height);
		result[3] = new Coordinate(x, y + height);
		result[4] = new Coordinate(x, y);
		return result;
	}

	/**
	 * Does this bounding box contain the given bounding box?
	 * 
	 * @param other
	 *            Another Bbox.
	 * @return true if the other is completely surrounded by this one, false otherwise.
	 */
	public boolean contains(Bbox other) {
		if (other.getX() < this.getX()) {
			return false;
		}
		if (other.getY() < this.getY()) {
			return false;
		}
		if (other.getEndPoint().getX() > this.getEndPoint().getX()) {
			return false;
		}
		if (other.getEndPoint().getY() > this.getEndPoint().getY()) {
			return false;
		}
		return true;
	}

	/**
	 * Does this bounding box intersect the given bounding box?
	 * 
	 * @param other
	 *            Another Bbox.
	 * @return true if the other intersects this one, false otherwise.
	 */
	public boolean intersects(Bbox other) {
		if (other.getX() > this.getEndPoint().getX()) {
			return false;
		}
		if (other.getY() > this.getEndPoint().getY()) {
			return false;
		}
		if (other.getEndPoint().getX() < this.getX()) {
			return false;
		}
		if (other.getEndPoint().getY() < this.getY()) {
			return false;
		}
		return true;
	}

	/**
	 * Computes the intersection of this bounding box with the specified bounding box.
	 * 
	 * @param other
	 *            Another Bbox.
	 * @return bounding box of intersection or null if they do not intersect.
	 */
	public Bbox intersection(Bbox other) {
		if (!this.intersects(other)) {
			return null;
		} else {
			double minx = other.getX() > this.getX() ? other.getX() : this.getX();
			double maxx = other.getEndPoint().getX() < this.getEndPoint().getX() ? other.getEndPoint().getX() : this
					.getEndPoint().getX();
			double miny = other.getY() > this.getY() ? other.getY() : this.getY();
			double maxy = other.getEndPoint().getY() < this.getEndPoint().getY() ? other.getEndPoint().getY() : this
					.getEndPoint().getY();
			return new Bbox(minx, miny, (maxx - minx), (maxy - miny));
		}
	}

	/**
	 * Calculates the union of 2 bounding boxes.
	 * 
	 * @param other
	 *            The other Bbox. Can be a bounding box with width and height equal to 0.
	 */
	public Bbox union(Bbox other) {
		if (other.getWidth() == 0 && other.getHeight() == 0 && other.getX() == 0 && other.getY() == 0) {
			return (Bbox) clone();
		}
		if (width == 0 && height == 0 && x == 0 && y == 0) {
			return (Bbox) other.clone();
		}

		double minx = other.getX() < this.getX() ? other.getX() : this.getX();
		double maxx = other.getEndPoint().getX() > this.getEndPoint().getX() ? other.getEndPoint().getX() : this
				.getEndPoint().getX();
		double miny = other.getY() < this.getY() ? other.getY() : this.getY();
		double maxy = other.getEndPoint().getY() > this.getEndPoint().getY() ? other.getEndPoint().getY() : this
				.getEndPoint().getY();
		return new Bbox(minx, miny, (maxx - minx), (maxy - miny));
	}

	/**
	 * Return a new bounding box that has increased in size by adding a range to this bounding box.
	 * 
	 * @param range
	 *            Must be a positive number, otherwise null will be returned.
	 * @return
	 */
	public Bbox buffer(double range) {
		if (range > 0) {
			double r2 = range * 2;
			return new Bbox(x - range, y - range, width + r2, height + r2);
		}
		return null;
	}

	/**
	 * Return a new bounding box which has the same center position but has been scaled with the specified factor.
	 * 
	 * @param factor
	 *            The scale factor (must be > 0).
	 * @return
	 */
	public Bbox scale(double factor) {
		if (factor > 0) {
			double scaledWidth = width * factor;
			double scaledHeight = height * factor;
			Coordinate center = getCenterPoint();
			return new Bbox(center.getX() - scaledWidth / 2, center.getY() - scaledHeight / 2, scaledWidth,
					scaledHeight);
		} else {
			return new Bbox(this);
		}
	}

	/**
	 * Translates this bounds with displacement dx and dy.
	 * 
	 * @param dx
	 *            x displacement
	 * @param dy
	 *            y displacement
	 */
	public void translate(double dx, double dy) {
		this.x = this.x + dx;
		this.y = this.y + dy;
	}

	/**
	 * Create a new bounds by transforming this bounds with the specified tranformation matrix.
	 * 
	 * @param t
	 *            the transformation matrix
	 * @return the transformed bounds
	 */
	public Bbox transform(Matrix t) {
		Coordinate c1 = transform(t, new Coordinate(x, y));
		Coordinate c2 = transform(t, new Coordinate(x + width, y + height));
		Coordinate origin = new Coordinate(Math.min(c1.getX(), c2.getX()), Math.min(c1.getY(), c2.getY()));
		Coordinate endPoint = new Coordinate(Math.max(c1.getX(), c2.getX()), Math.max(c1.getY(), c2.getY()));
		return new Bbox(origin.getX(), origin.getY(), endPoint.getX() - origin.getX(), endPoint.getY() - origin.getY());
	}

	/**
	 * Moves center to the specified coordinate.
	 * 
	 * @param coordinate
	 *            new center point
	 */
	public void setCenterPoint(Coordinate coordinate) {
		this.x = coordinate.getX() - 0.5 * this.width;
		this.y = coordinate.getY() - 0.5 * this.height;
	}

	/**
	 * Returns whether or not this bounding box is empty. A bounding box is considered empty when either the width or
	 * the height is equal to zero.
	 */
	public boolean isEmpty() {
		return width == 0 || height == 0;
	}

	/**
	 * Return a nice print of this bounding box.
	 */
	public String toString() {
		return "Bbox[" + x + " " + y + ", " + width + " " + height + "]";
	}

	// -------------------------------------------------------------------------
	// Getters and setters:
	// -------------------------------------------------------------------------

	public double getHeight() {
		return height;
	}

	public void setHeight(double height) {
		this.height = height;
	}

	public double getWidth() {
		return width;
	}

	public void setWidth(double width) {
		this.width = width;
	}

	public double getX() {
		return x;
	}

	public void setX(double x) {
		this.x = x;
	}

	public double getY() {
		return y;
	}

	public void setY(double y) {
		this.y = y;
	}

	public double getMaxX() {
		return getX() + getWidth();
	}

	public double getMaxY() {
		return getY() + getHeight();
	}

	public boolean equals(Bbox other, double delta) {
		return equals(this.x, other.x, delta) && equals(this.y, other.y, delta)
				&& equals(this.width, other.width, delta) && equals(this.height, other.height, delta);
	}

	protected boolean equals(double d1, double d2, double delta) {
		return Math.abs(d1 - d2) <= delta;
	}

	private Coordinate transform(Matrix t, Coordinate coordinate) {
		double x = t.getXx() * coordinate.getX() + t.getXy() * coordinate.getY() + t.getDx();
		double y = t.getYx() * coordinate.getY() + t.getYy() * coordinate.getY() + t.getDy();
		return new Coordinate(x, y);
	}
}